Userland/Utilities/tac.cpp

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright (c) 2021, Federico Guerinoni <guerinoni.federico@gmail.com>
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <AK/Vector.h>
+#include <LibCore/ArgsParser.h>
+#include <LibCore/File.h>
+#include <unistd.h>
+
+void print_lines(const Vector<String>& vec);
+
+int main(int argc, char** argv)
+{
+    if (pledge("stdio rpath", nullptr) < 0) {
+        perror("pledge");
+        return 1;
+    }
+
+    Vector<const char*> paths;
+
+    Core::ArgsParser args_parser;
+    args_parser.set_general_help("Concatenate files or pipes to stdout, last line first.");
+    args_parser.add_positional_argument(paths, "File path", "path", Core::ArgsParser::Required::No);
+    args_parser.parse(argc, argv);
+
+    Vector<String> lines;
+    if (paths.is_empty()) {
+        char buf[32768];
+        ssize_t nread = read(0, buf, sizeof(buf));
+        if (nread == 0)
+            return 1;
+        if (nread < 0) {
+            perror("read");
+            return 2;
+        }
+
+        auto vec = String(buf, Chomp).split('\n', true);
+        for (auto&& s : vec) {
+            lines.append(move(s));
+        }
```

Comment:
This code still does not make sense. What if there's more than 32768 bytes coming on  `stdin`? We will not process any more than that.

Instead of having two separate code paths for `paths.is_empty()` being true/false, I suggest doing everything on one code path, and simply using `Core::File::standard_input()` in the stdin case, which gives you a `Core::File` wrapper around stdin.